File: Php/app/Cashgame/Action/GameChartData.php
namespace app\Cashgame\Action {

	use DateTime;
	use app\Chart\ChartDateTimeColumnModel;
	use app\Chart\ChartDateTimeValueModel;
	use app\Chart\ChartModel;
	use app\Chart\ChartNumberColumnModel;
	use app\Chart\ChartRowModel;
	use app\Chart\ChartValueModel;
	use core\DateTimeFactory;
	use entities\Cashgame;
	use entities\CashgameResult;
	use entities\Checkpoints\CheckpointType;
	use entities\GameStatus;
	use entities\Homegame;

	class GameChartData extends ChartModel{

		private $homegame;
		private $cashgame;

		public function __construct(Homegame $homegame, Cashgame $cashgame){
			homegame = $homegame;
			cashgame = $cashgame;
			addActionColumns();
			addActionRows();
		}

		private function addActionRows(){
			$results = cashgame.getResults();
			foreach($results as $result){
				$totalBuyin = 0;
				$checkpoints = $result.getCheckpoints();
				$playerName = $result.getPlayer().getDisplayName();
				foreach($checkpoints as $checkpoint){
					if($checkpoint.getType() == CheckpointType::buyin){
						$totalBuyin += $checkpoint.getAmount();
					}
					addActionRow($checkpoint.getTimestamp(), $checkpoint.getStack() - $totalBuyin, $playerName);
				}
			}
			if(cashgame.getStatus() == GameStatus::running){
				addCurrentStacks($results);
			}
		}

		private function addCurrentStacks($results){
			$timestamp = DateTimeFactory::now(homegame.getTimezone());
			$row = new ChartRowModel();
			$row.addValue(new ChartDateTimeValueModel($timestamp));
			foreach($results as $result){  /** @var $result CashgameResult */
				$winnings = $result.getStack() - $result.getBuyin();
				$row.addValue(new ChartValueModel($winnings));
			}
			addRow($row);
		}

		private function addActionColumns(){
			$timeCol = new ChartDateTimeColumnModel('Time', 'HH:mm');
			addColumn($timeCol);
			$playerNames = cashgame.getPlayerNames();
			foreach($playerNames as $playerName){
				$playerCol = new ChartNumberColumnModel($playerName);
				addColumn($playerCol);
			}
		}

		private function addActionRow(DateTime $dateTime, $winnings, $currentPlayerName){
			$row1 = new ChartRowModel();
			$row1.addValue(new ChartDateTimeValueModel($dateTime));
			$playerNames = cashgame.getPlayerNames();
			foreach($playerNames as $playerName){
				$val = null;
				if($playerName == $currentPlayerName){
					$val = $winnings;
				}
				$row1.addValue(new ChartValueModel($val));
			}
			addRow($row1);
		}

	}

}

File: Web/Views/Cashgame/Chart/GameChartData.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Core.Classes;
using Core.Classes.Checkpoints;
using Infrastructure.System;
using Web.Models.ChartModels;

namespace Web.Views.Cashgame.Chart {

	public class GameChartData : ChartModel{

		public GameChartData(Homegame homegame, Core.Classes.Cashgame cashgame){
			AddActionColumns(cashgame);
			AddActionRows(homegame, cashgame);
		}

		private void AddActionRows(Homegame homegame, Core.Classes.Cashgame cashgame){
			var results = cashgame.Results;
			foreach(var result in results){
				var totalBuyin = 0;
				var checkpoints = result.Checkpoints;
				var playerName = result.Player.DisplayName;
				foreach(var checkpoint in checkpoints){
					if(checkpoint.Type == CheckpointType.Buyin){
						totalBuyin += checkpoint.Amount;
					}
					AddActionRow(cashgame, checkpoint.Timestamp, checkpoint.Stack - totalBuyin, playerName);
				}
			}
			if(cashgame.Status == GameStatus.Running){
				AddCurrentStacks(homegame, results);
			}
		}

		private void AddCurrentStacks(Homegame homegame, IEnumerable<CashgameResult> results){
			var timestamp = DateTimeFactory.Now(homegame.Timezone);
			var row = new ChartRowModel();
			row.AddValue(new ChartDateTimeValueModel(timestamp));
			foreach(var result in results){
				var winnings = result.Stack - result.Buyin;
				row.AddValue(new ChartValueModel(winnings));
			}
			AddRow(row);
		}

		private void AddActionColumns(Core.Classes.Cashgame cashgame){
			AddColumn(new ChartDateTimeColumnModel("Time", "HH:mm"));
			var playerNames = cashgame.GetPlayerNames();
			foreach(var playerName in playerNames){
				AddColumn(new ChartNumberColumnModel(playerName));
			}
		}

		private void AddActionRow(Core.Classes.Cashgame cashgame, DateTime dateTime, int winnings, string currentPlayerName){
			var row1 = new ChartRowModel();
			row1.AddValue(new ChartDateTimeValueModel(dateTime));
			var playerNames = cashgame.GetPlayerNames();
			foreach(var playerName in playerNames){
				string val = null;
				if(playerName == currentPlayerName){
					val = winnings.ToString(CultureInfo.InvariantCulture);
				}
				row1.AddValue(new ChartValueModel(val));
			}
			AddRow(row1);
		}

	}

}